bpo-39622: Interrupt the main asyncio task on Ctrl+C (#32105)
python/cpython: Doc/library/asyncio-runner.rst

      Embedded *loop* and *context* are created at the :keyword:`with` body entering
      or the first call of :meth:`run` or :meth:`get_loop`.

> Handling Keyboard Interruption

Author: @kumaraditya303 could you review this section, please?
Feel free to suggest wild rephrasing if you don't like the chapter structure.

Reviewer: Sure, I have added inline suggestions.

==============================

.. versionadded:: 3.11

> When a user hits the interrupt key (normally :kbd:`Ctrl-C`), :exc:`KeyboardInterrupt`

Reviewer: ```suggestion
When :const:`signal.SIGINT` is raised by :kbd:`Ctrl-C`, :exc:`KeyboardInterrupt`
```

> exception is raised in the main thread by default.

Reviewer: ```suggestion
exception is raised in the main thread by default. However this doesn't work with :mod:`asyncio` 
because it can interrupt asyncio internals and can hang the program from exiting.

To mitigate this issue, :mod:`asyncio` handles :const:`signal.SIGINT` as follows:
```

> This approach doesn't work with asyncio, the interruption should never break asyncio
> internals itself but stop a user code.
> 
> Asyncio handles the keyboard interruption as follows:

Reviewer: ```suggestion

```

> 1. :meth:`asyncio.Runner.run` installs a custom :const:`signal.SIGINT` handler before
>    async function execution and reverts it back at the exit from the function.

Reviewer: ```suggestion
1. :meth:`asyncio.Runner.run` installs a custom :const:`signal.SIGINT` handler before
   any user code is executed and removes it when exiting from the function.
```

> 2. The :class:`~asyncio.Runner` creates the main task for passed async function
>    processing.

Reviewer: ```suggestion
2. The :class:`~asyncio.Runner` creates the main task for the passed coroutine for 
    its execution.
```

> 3. When the interruption is requested, the custom signal handler cancels the main task
>    by :meth:`~asyncio.Task.cancel` call.  :exc:`asyncio.CancelledError` is raised
>    inside the the main task.  The exception unwinds, ``try/except`` and ``try/finally``
>    blocks can be used for resource cleanup.  After the main task is finished,
>    :meth:`asyncio.Runner.run` raises :exc:`KeyboardInterrupt`.

Reviewer: ```suggestion
3. When :const:`signal.SIGINT` is raised by :kbd:`Ctrl-C`, the custom signal handler
   cancels the main task by calling :meth:`asyncio.Task.cancel` which raises 
   :exc:`asyncio.CancelledError` inside the the main task.  This causes the Python stack 
   to unwind,  ``try/except`` and ``try/finally`` blocks can be used for resource cleanup. 
   After the main task is cancelled, :meth:`asyncio.Runner.run` raises :exc:`KeyboardInterrupt`.
```

> 4. A user could write a thight loop, :meth:`~asyncio.Task.cancel` cannot break it.
>    That's why the second and consequent :kbd:`Ctrl-C` hits immediatelly raise
>    :exc:`KeyboardInterrupt` to stop the program abnormally.

Reviewer: ```suggestion
4. A user could write a tight loop which cannot be interrupted by :meth:`asyncio.Task.cancel`,
   in which case the second following :kbd:`Ctrl-C` immediately raises the :exc:`KeyboardInterrupt`
   without cancelling the main task.
```
